Nodejs/Product/Nodejs/SharedProject/Wpf/NotificationTextBox.cs

```diff
@@ -0,0 +1,103 @@
+{
+    internal class NotificationTextBox : TextBox
+    {
+        // This control's AutomationPeer is the object that actually raises the UIA Notification event.
+        private NotificationTextBoxAutomationPeer notificationTextBoxAutomationPeer;
+
+        // Assume the UIA Notification event is available until we learn otherwise.
+        // If we learn that the UIA Notification event is not available, no instance
+        // of the NotificationTextBox should attempt to raise it.
+        private static bool notificationEventAvailable = true;
+        public bool NotificationEventAvailable
+        {
+            get
+            {
+                return notificationEventAvailable;
+            }
+            set
+            {
+                notificationEventAvailable = value;
+            }
+        }
+
+        protected override AutomationPeer OnCreateAutomationPeer()
+        {
+            this.notificationTextBoxAutomationPeer = new NotificationTextBoxAutomationPeer(this);
+
+            return this.notificationTextBoxAutomationPeer;
+        }
+
+        public void RaiseNotificationEvent(string notificationText, string notificationGuid)
+        {
+            // Only attempt to raise the event if we already have an AutomationPeer.
+            if (this.notificationTextBoxAutomationPeer != null)
+            {
+                this.notificationTextBoxAutomationPeer.RaiseNotificationEvent(notificationText, notificationGuid);
+            }
+        }
+    }
+
+    internal class NotificationTextBoxAutomationPeer : TextBoxAutomationPeer
+    {
+        private NotificationTextBox notificationTextBox;
+
+        // The UIA Notification event requires the IRawElementProviderSimple
+        // associated with this AutomationPeer.
+        private IRawElementProviderSimple rawElementProviderSimple;
+
+        public NotificationTextBoxAutomationPeer(NotificationTextBox owner) : base(owner)
+        {
+            this.notificationTextBox = owner;
+        }
+
+        public void RaiseNotificationEvent(string notificationText, string notificationGuid)
+        {
+            // If we already know that the UIA Notification event is not available, do not
+            // attempt to raise it.
+            if (this.notificationTextBox.NotificationEventAvailable)
+            {
+                // If no UIA clients are listening for events, don't bother raising one.
+                if (UiaAutomationNativeMethods.UiaClientsAreListening())
+                {
+                    // Get the IRawElementProviderSimple for this AutomationPeer if we don't
+                    // have it already.
+                    if (this.rawElementProviderSimple == null)
+                    {
+                        var automationPeer = FromElement(this.notificationTextBox);
+                        if (automationPeer != null)
+                        {
+                            this.rawElementProviderSimple = this.ProviderFromPeer(automationPeer);
```

Comment:
Is this expensive?  Why cache it?

Reply:
Looking through the source it does appears that in order to get the provider it might navigate the UIA tree, which I'm not entirely sure what it means but sounds expensive.